WeatherStationClient.py

```diff
@@ -39,79 +1063,231 @@ def __init__(self, client):
+
+            async def send(*args):
+                try:
+                    ret = await self._send(*args)
+                    if type == Model and isinstance(ret.response, dict):
+                        return getattr(model._from_response(ret), factory_name)(Model._dict(ret.response))
+                    elif type == list and isinstance(ret.response, list):
+                        return ListResponse._from_response(ret)\
+                            (Model._from_list(ret.response,
+                                              {'factory': getattr(model, factory_name), 'pass_index': pass_index}))
+                    elif type == bool and isinstance(ret.response, bool):
+                        return ApiBoolResponse._from_response(ret)(ret.response)
+                    elif type == dict and isinstance(ret.response, dict):
+                        return DictResponse._from_response(ret)\
+                            (Model._from_dict(ret.response,
+                                              {'factory': getattr(model, factory_name), 'pass_index': pass_index}))
+                    else:
+                        return ret
+                except ApiResponseException as api_response_exception:
+                    raise api_response_exception
+            return send
+
+        @staticmethod
+        def _from_model(model, factory_name):
+            try:
+                factory = getattr(model, factory_name)
+            except AttributeError:
+                pass
+            if 'factory' in locals():
+                return factory()
+            else:
+                return model
+
+    class Shortcuts(ClientRoute):
+        """
+        The methods of this class do not have a 1-1 correspondence to any particualr official API routes.
+        Instead, these are convenience methods, wrapping up in one call what would normally require the user to make
+        multiple calls.
+        """
+
+        async def get_stations_by_types(self):
+            """
+            Returns stations attached to a user account already grouped by station types.
+            Is equivalent to the following code:
+                stations = await client.user.get_stations()
+                types = await client.system.get_station_types()
+                for station_type in types.values():
+                    station_type.bind(stations)
+            """
+
+            stations, types = await gather(self._client.user.get_stations(), self._client.system.get_types())
+
+            if isinstance(types, dict):
+                for station_type in types.values():
+                    if isinstance(station_type, StationType):
+                        station_type.bind(stations)
+
+            return types
 
-        async def user_information(self):
-            """Reading user information."""
-            return await self._send('GET', 'user')
+    class User(ClientRoute):
```

Comment:
Maybe its better idea to set modifications to api slide. Two reasons:
1) its alraedy tested
2) we can build full fledged system consisted of "independanet" tiers 
tcp/ip we dont need to break ethernet to guarantee multi network communication

In some cases we could event let user to do translation json <-> model himself, if we define serialization, deserialization from/to json its one liner.

Reply:
Nie do końca rozumiem, o co Ci chodzi :( Czy o to, żeby `client.user.get()` nie zwracała modelu?

>     1. its alraedy tested

No dobrze ale deserializację ze słownika też trzeba przetestować. Może jeszcze nie widzę problemu, który niedługo stanie się oczywisty, ale chyba, aby przetestować to, co robię, wystarczy zamockować wartości zwracane przez `_send`, co i tak robimy?

>     2\. we can build full fledged system consisted of "independanet" tiers
        tcp/ip we dont need to break ethernet to guarantee multi network communication

Bardzo się starałem, żeby umożliwić właśnie to, o czym piszesz. Metoda `user.get()` i jej podobne zwracają obiekt, który dziedziczy zarówno po modelu `User`, jak i po `ApiResponse`. W konsekwencji user może, jeśli chce, albo zapomnieć o tym, że są to modele, i operować na czystych wartościach, zwracanych przez serwer (zdeserializowanych już przez `aiohttp`, nie przeze mnie, z JSONa do słownika), albo odwrotnie: zapomnieć o niższym poziomie i operować tylko na modelach.

> In some cases we could event let user to do translation json <-> model himself, if we define serialization, deserialization from/to json its one liner.

Ale może, jeśli chce!

Mamy metodę `User._from_get` - jeśli user chce, może ja sobie wywołać ręcznie. Może też ją sobie zoverride'ować, jeśli zrobi własną podklasę User (chociaż wtedy rzeczywiście będzie musiał ją wołać ręcznie). Może wreszcie o niej zapomnieć i budować sobie model User samodzielnie, przekazując konstruktorowi odpowiednie parametry.

Może nawet należałoby upublicznić tę metodę, usuwając `_` z początku jej nazwy. Ale to drobiazg.

Reply:
Po zastanowieniu: Nie wiem. Być może masz rację.

To na przykład, że uparłem się, by zwracać obiekty dziedziczące zarówno po ApiResponse jak i po Model: jak widać, jest do zrobienia, jednak zapewnienie tego to jest robota i odbywa się to kosztem wprowadzania tego meta-meta-meta. Chociaż dla usera może być to wygodne

Albo dalej: naprawianie referencji. Stacja ma swój typ, typ to też model: uff. Z perspektywy usera pewnie wygodne byłoby, gdyby stacja nie miała pola z ID typu, tylko referencję do typu. Co można zapewnić. Ale teraz mamy metodę `station.get`, która zwraca nowy obiekt typu station. NO i co? user sam ma naprawiać referencje? No to już się upierdliwe może zrobić. Z drugiej strony, jeśli każda metoda ma naprawiać referencje, to... copypasty. Kroją nam się kolejne meta-meta-meta, jeśli chcemy uniknąć tychże copypast

Być może stoimy przed takim wyborem: albo pozwolimy userowi (jak oryginalnie chciałem) rzeczywiście zapomnieć, że serwer zwraca słowniki, i pozwolić mu (o ile chce i bez odbierania możliwości pracy na słownikach) pracować wyłącznie na modelach - co może być dla usera wygodne - ale koszt tego to większa komplikacja kodu oraz więcej pracy dla nas; albo zamiast tego wyraźnie wymusimy pracę także na "czystych" odpowiedziach serwera, co - być może - może stać się dla usera upierdliwe, ale my będziemy mieć mniej pracy i kod będzie prostszy.

Reply:
> Mamy metodę `User._from_get` - jeśli user chce, może ja sobie wywołać ręcznie. Może też ją sobie zoverride'ować, jeśli zrobi własną podklasę User (chociaż wtedy rzeczywiście będzie musiał ją wołać ręcznie).

Jeśli jednak chcemy iść tym ambitniejszym(?) torem: w zasadzie nawet nie będzie musiał jej wołać ręcznie.

Niech metoda `user.get()` przyjmuje dodatkowy, opcjonalny argument, tj klasę, której instancję zbuduje. Z kolei te wszystkie meta-meta-meta powinny uprościć userowi zbudowanie własnej metody `_from_get` tak, żeby nie rzucała ona wyjątkami przy każdej niezgodności odpowiedzi serwera z dokumentacją

Reply:
Taka funkcjonalność, choć być może przydatna, znajduje się jednak chyba najniżej listy priorytetów